File: packages/deon-javascript/scripts/adapter.mjs
// The cross-implementation harness adapter (`spec/harness/README.md`).
//
// A filter: newline-delimited JSON in, newline-delimited JSON out. Nothing escapes it but a
// response — a host exception crossing this boundary would be reported as a disagreement, and it
// would be one.

import { createInterface } from 'node:readline';

import Deon, { typer } from '../distribution/index.mjs';


const flag = (request, name, fallback = false) =>
    (request[name] ?? (fallback ? 'true' : 'false')) === 'true';


const optionsOf = (request) => ({
    sourceName: request.sourceName ?? '<memory>',
    filebase: request.filebase ?? '',
    resources: request.files ?? {},
    absolutePaths: request.absolutePaths ?? {},
    environment: request.environment ?? {},
    allowFilesystem: flag(request, 'allowFilesystem'),
    allowNetwork: flag(request, 'allowNetwork'),

    // The contracts of specification 14.1. They arrive through `files` like every other resource, so
    // no adapter reaches a disk.
    datasignFiles: request.datasignFiles ?? [],
    datasignMap: request.datasignMap ?? {},
});


const stringifyOptionsOf = (given = {}) => ({
    canonical: given.canonical === 'true',
    readable: (given.readable ?? 'true') === 'true',
    indentation: Number(given.indentation ?? '4'),
    leaflinks: given.leaflinks === 'true',
    leaflinkLevel: Number(given.leaflinkLevel ?? '1'),
    leaflinkShortening: (given.leaflinkShortening ?? 'true') === 'true',
    generatedHeader: given.generatedHeader === 'true',
    generatedComments: given.generatedComments === 'true',
});


const run = (request) => {
    const deon = new Deon();
    const options = optionsOf(request);
    const { op, source } = request;

    if (op === 'entities') {
        return JSON.stringify(
            deon.entities(source, options.sourceName).map(entity => ({
                name: entity.name,
                parameters: entity.parameters,
                kind: entity.kind,
            })),
        );
    }

    if (op === 'lint') {
        return JSON.stringify(
            deon.lint(source, options.sourceName).map(diagnostic => ({
                code: diagnostic.code,
                line: String(diagnostic.range.start.line),
                column: String(diagnostic.range.start.column),
            })),
        );
    }

    const value = deon.parseSynchronous(source, options);

    if (op === 'canonical') {
        return deon.canonical(value);
    }

    if (op === 'stringify') {
        return deon.stringify(value, stringifyOptionsOf(request.stringifyOptions));
    }

    if (op === 'typed') {
        return JSON.stringify(typer(value));
    }

    if (op === 'datasign') {
        // `parseSynchronous` has already applied the contracts.
        return JSON.stringify(value);
    }

    throw new Error(`unknown operation '${op}'`);
};


const lines = createInterface({ input: process.stdin, crlfDelay: Infinity });

for await (const line of lines) {
    if (!line.trim()) {
        continue;
    }

    const request = JSON.parse(line);

    let answer;

    try {
        answer = { id: request.id, ok: 'true', result: run(request) };
    } catch (failure) {
        const diagnostic = failure?.diagnostics?.[0];

        answer = diagnostic
            ? {
                id: request.id,
                ok: 'false',
                code: failure.code,
                line: String(diagnostic.range.start.line),
                column: String(diagnostic.range.start.column),
            }
            : {
                id: request.id,
                ok: 'false',
                code: `HOST_${failure?.constructor?.name ?? 'Error'}`,
                line: '0',
                column: '0',
            };
    }

    process.stdout.write(JSON.stringify(answer) + '\n');
}

File: packages/deon-python/harness/adapter.py
"""The cross-implementation harness adapter (`spec/harness/README.md`).

A filter: newline-delimited JSON in, newline-delimited JSON out. Nothing escapes it but a response —
a host exception crossing this boundary would be reported as a disagreement, and it would be one.
"""

from __future__ import annotations

import json
import sys

import deon
from deon import DeonError, ParseOptions, StringifyOptions


def flag(request: dict, name: str, fallback: bool = False) -> bool:
    return request.get(name, "true" if fallback else "false") == "true"


def options_of(request: dict) -> ParseOptions:
    options = ParseOptions()

    options.source_name = request.get("sourceName", "<memory>")
    options.filebase = request.get("filebase", "")
    options.resources = dict(request.get("files") or {})
    options.absolute_paths = dict(request.get("absolutePaths") or {})
    options.environment = dict(request.get("environment") or {})
    options.allow_filesystem = flag(request, "allowFilesystem")
    options.allow_network = flag(request, "allowNetwork")

    # A configurable resource budget arrives as a string count under `budgets` (specification 11).
    # Absent or 0 leaves the evaluator on its default expansion limit.
    budgets = request.get("budgets") or {}
    options.expansion = int(budgets.get("expansion") or 0)

    # The contracts of specification 14.1. They arrive through `files` like every other resource, so
    # no adapter reaches a disk.
    options.datasign_files = list(request.get("datasignFiles") or [])
    options.datasign_map = dict(request.get("datasignMap") or {})

    return options


def stringify_options_of(given: dict) -> StringifyOptions:
    return StringifyOptions(
        canonical=given.get("canonical") == "true",
        readable=given.get("readable", "true") == "true",
        indentation=int(given.get("indentation", "4")),
        leaflinks=given.get("leaflinks") == "true",
        leaflink_level=int(given.get("leaflinkLevel", "1")),
        leaflink_shortening=given.get("leaflinkShortening", "true") == "true",
        generated_header=given.get("generatedHeader") == "true",
        generated_comments=given.get("generatedComments") == "true",
    )


def run(request: dict) -> str:
    operation = request["op"]
    source = request["source"]
    options = options_of(request)

    if operation == "entities":
        found = deon.entities(source, options.source_name)

        return json.dumps(
            [
                {"name": e.name, "parameters": list(e.parameters), "kind": e.kind}
                for e in found
            ]
        )

    if operation == "lint":
        return json.dumps(
            [
                {"code": d.code, "line": str(d.line), "column": str(d.column)}
                for d in deon.lint(source, options.source_name)
            ]
        )

    value = deon.parse_with(source, options)

    if operation == "canonical":
        return deon.canonical(value)

    if operation == "stringify":
        return deon.stringify(value, stringify_options_of(request.get("stringifyOptions") or {}))

    if operation == "typed":
        return json.dumps(deon.typed(value))

    if operation == "datasign":
        # `parse_with` has already applied the contracts, as the reference implementation does.
        return json.dumps(value)

    raise ValueError(f"unknown operation '{operation}'")


def main() -> None:
    for line in sys.stdin:
        line = line.strip()

        if not line:
            continue

        request = json.loads(line)

        try:
            answer = {"id": request["id"], "ok": "true", "result": run(request)}
        except DeonError as failure:
            span = failure.diagnostics[0].span

            answer = {
                "id": request["id"],
                "ok": "false",
                "code": failure.code,
                "severity": failure.diagnostics[0].severity,
                "start": str(span.start),
                "line": str(span.line),
                "column": str(span.column),
                "related": [
                    [str(s.start), str(s.line), str(s.column)] for s in failure.diagnostics[0].related
                ],
            }
        except Exception as failure:  # the host leaking through is a disagreement, and it says so
            answer = {
                "id": request["id"],
                "ok": "false",
                "code": f"HOST_{type(failure).__name__}",
                "line": "0",
                "column": "0",
            }

        sys.stdout.write(json.dumps(answer) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()
